crates/fuel-core/src/database/contracts.rs

```diff
@@ -16,19 +16,114 @@ use fuel_core_storage::{
         ContractsRawCode,
     },
     ContractsAssetKey,
+    Error as StorageError,
+    Mappable,
     Result as StorageResult,
     StorageAsRef,
+    StorageInspect,
+    StorageMutate,
 };
-use fuel_core_types::fuel_types::{
-    AssetId,
-    Bytes32,
-    ContractId,
-    Word,
+use fuel_core_types::{
+    fuel_tx::Contract,
+    fuel_types::{
+        AssetId,
+        Bytes32,
+        ContractId,
+        Word,
+    },
 };
+use serde::Deserializer;
+use std::borrow::Cow;
 
-impl DatabaseColumn for ContractsRawCode {
-    fn column() -> Column {
-        Column::ContractsRawCode
+struct OptimizedContract {
+    bytes: Vec<u8>,
+}
+
+impl From<OptimizedContract> for Contract {
+    fn from(value: OptimizedContract) -> Self {
+        value.bytes.into()
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for OptimizedContract {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct BytesVisitor;
+
+        impl<'a> serde::de::Visitor<'a> for BytesVisitor {
+            type Value = Vec<u8>;
+
+            fn expecting(
+                &self,
+                formatter: &mut core::fmt::Formatter,
+            ) -> core::fmt::Result {
+                formatter.write_str("a borrowed byte array")
+            }
+
+            fn visit_borrowed_bytes<E>(self, v: &'a [u8]) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(v.to_vec())
+            }
+        }
+
+        let bytes = deserializer.deserialize_bytes(BytesVisitor {})?;
+        Ok(Self { bytes })
+    }
+}
+
+impl StorageInspect<ContractsRawCode> for Database {
+    type Error = StorageError;
+
+    fn get(
+        &self,
+        key: &<ContractsRawCode as Mappable>::Key,
+    ) -> Result<Option<Cow<<ContractsRawCode as Mappable>::OwnedValue>>, Self::Error>
+    {
+        self.get::<OptimizedContract>(key.as_ref(), Column::ContractsRawCode)
+            .map(|option| option.map(|contract| Cow::Owned(Contract::from(contract))))
+            .map_err(Into::into)
+    }
+
+    fn contains_key(
+        &self,
+        key: &<ContractsRawCode as Mappable>::Key,
+    ) -> Result<bool, Self::Error> {
+        self.contains_key(key.as_ref(), Column::ContractsRawCode)
+            .map_err(Into::into)
+    }
+}
+
+impl StorageMutate<ContractsRawCode> for Database {
+    fn insert(
+        &mut self,
+        key: &<ContractsRawCode as Mappable>::Key,
+        value: &<ContractsRawCode as Mappable>::Value,
+    ) -> Result<Option<<ContractsRawCode as Mappable>::OwnedValue>, Self::Error> {
+        Database::insert::<_, _, OptimizedContract>(
+            self,
+            key.as_ref(),
+            Column::ContractsRawCode,
+            &value,
+        )
+        .map(|option| option.map(Into::into))
+        .map_err(Into::into)
+    }
```

Comment:
On master this insert adds the contract in a way that it can be loaded directly into vm memory and executed. Is this also the case here?

Reply:
Yes, it is what we want to do, and also improve the performance of the `contracts` endpoint